File: day7/main.jl
#!/usr/bin/env julia

using DataStructures: Queue, enqueue!, dequeue!

function load_grid(lines)
    grid = String[]
    sr = -1; sc = -1
    for (r,line) in enumerate(lines)
        push!(grid, line)
        pos = findfirst(==('S'), line)
        if pos !== nothing
            sr = r; sc = pos
        end
    end
    sr < 0 && error("missing S")
    return grid, sr, sc
end

function part1(grid, sr, sc)
    h = length(grid); w = lastindex(grid[1])
    active = Set{Int}([sc])
    splits = 0
    for r in sr:h
        next = Set{Int}()
        seen = Set{Int}()
        queue = Queue{Int}()
        for c in active
            enqueue!(queue, c)
        end
        while !isempty(queue)
            c = dequeue!(queue)
            c in seen && continue
            push!(seen, c)
            cell = grid[r][c]
            if cell == '^'
                splits += 1
                c > 1 && enqueue!(queue, c-1)
                c < w && enqueue!(queue, c+1)
            else
                push!(next, c)
            end
        end
        active = next
        isempty(active) && break
    end
    splits
end

function part2(grid, sr, sc)
    h = length(grid); w = lastindex(grid[1])
    active = Dict{Int,Int}(sc => 1)
    for r in sr:h
        next = Dict{Int,Int}()
        for (c,count) in active
            cell = grid[r][c]
            if cell == '^'
                c > 1 && (next[c-1] = get(next,c-1,0) + count)
                c < w && (next[c+1] = get(next,c+1,0) + count)
            else
                next[c] = get(next,c,0) + count
            end
        end
        active = next
        isempty(active) && break
    end
    sum(values(active))
end

function main()
    t0 = time_ns()
    lines = readlines("input.txt")
    grid, sr, sc = load_grid(lines)
    p1 = part1(grid, sr, sc)
    p2 = part2(grid, sr, sc)
    elapsed_ms = (time_ns() - t0)/1e6
    println("splits=$(p1) timelines=$(p2) elapsed_ms=$(round(elapsed_ms; digits=3))")
end

main()

File: day7/Main.hs
{-# LANGUAGE BangPatterns #-}
module Main where

import Data.Maybe (fromJust)
import qualified Data.Set as S
import qualified Data.Map.Strict as M
import qualified Data.Sequence as Seq
import GHC.Clock (getMonotonicTimeNSec)

loadGrid :: [String] -> ([String], Int, Int)
loadGrid ls =
  let grid = ls
      sr = fromJust (findIndexElem 'S' grid)
      sc = case findIndex (=='S') (grid !! sr) of
             Just c -> c
             Nothing -> error "no S"
  in (grid, sr, sc)

findIndexElem :: Char -> [String] -> Maybe Int
findIndexElem ch = go 0
  where
    go _ [] = Nothing
    go i (x:xs) = if ch `elem` x then Just i else go (i+1) xs

findIndex :: (a -> Bool) -> [a] -> Maybe Int
findIndex p = go 0
  where
    go _ [] = Nothing
    go i (x:xs) = if p x then Just i else go (i+1) xs

part1 :: [String] -> Int -> Int -> Int
part1 grid sr sc =
  let h = length grid
      w = length (head grid)
      go r splits active
        | r >= h = splits
        | S.null active = splits
        | otherwise =
            let (next,s') = bfs r splits (Seq.fromList (S.toList active)) S.empty S.empty
            in go (r+1) s' next
      bfs r splits queue seen next
        | Seq.null queue = (next, splits)
        | otherwise =
            let (c Seq.:< rest) = Seq.viewl queue
            in if S.member c seen
                 then bfs r splits rest seen next
                 else let seen' = S.insert c seen
                          cell = grid !! r !! c
                      in if cell == '^'
                           then let splits' = splits + 1
                                    rest' = (if c > 0 then rest Seq.|> (c-1) else rest)
                                    rest''= if c+1 < w then rest' Seq.|> (c+1) else rest'
                                in bfs r splits' rest'' seen' next
                           else bfs r splits rest seen' (S.insert c next)
  in go sr 0 (S.singleton sc)

part2 :: [String] -> Int -> Int -> Int
part2 grid sr sc =
  let h = length grid
      w = length (head grid)
      go r active
        | r >= h = active
        | M.null active = active
        | otherwise = go (r+1) (step r active)
      step r act =
        M.foldlWithKey'
          (\acc c cnt ->
             let cell = grid !! r !! c
             in if cell == '^'
                  then let acc' = if c > 0 then M.insertWith (+) (c-1) cnt acc else acc
                           acc''= if c+1 < w then M.insertWith (+) (c+1) cnt acc' else acc'
                       in acc''
                  else M.insertWith (+) c cnt acc)
          M.empty act
      final = go sr (M.singleton sc 1)
  in sum (M.elems final)

main :: IO ()
main = do
  t0 <- getMonotonicTimeNSec
  linesIn <- lines <$> readFile "input.txt"
  let (grid, sr, sc) = loadGrid linesIn
      !p1 = part1 grid sr sc
      !p2 = part2 grid sr sc
  p1 `seq` p2 `seq` return ()
  t1 <- getMonotonicTimeNSec
  let elapsedMs = fromIntegral (t1 - t0) / 1e6 :: Double
  putStrLn $ "splits=" ++ show p1 ++ " timelines=" ++ show p2
             ++ " elapsed_ms=" ++ showFF elapsedMs
  where
    showFF x = let s = show (fromIntegral (round (x * 1000 :: Double)) / 1000 :: Double)
               in if '.' `elem` s then s else s ++ ".0"
